DOCS-11843-2.4-indexed-readers-kt (#40)
mulesoft/docs-dataweave: modules/ROOT/pages/dataweave-indexed-readers.adoc

> = Indexed Readers in DataWeave

Reviewer: Is this topic really about indexed readers or is it about the various methods that DataWeave uses to process input data that is too large for memory?

> In DataWeave some readers of input data (such as XML, JSON, and CSV) support an indexed strategy to avoid loading the whole document in memory while still allowing random access just like in the in-memory strategy. Indexed readers are used when the input is larger than a certain configurable threshold or if the `indexedReader` setting is set to `true`.

Reviewer 2: 1. In DataWeave some readers > In DataWeave**,** some readers...
2. Indexed readers are used > DataWeave uses indexed readers (confirm if this is true, because it seems that DW automatically uses indexed readers when the input is larger than the limit according to this)

To manage files that don’t fit in memory, write the files to a temporary file in disk and index the document content beforehand, which also takes time and disk space.

> Indexed readers can process input files up to 20GBs. For bigger files use the streaming mode, which is faster. Despite this mode being more limited in functionality, there’s no maximum input size limitation for streaming readers.

Reviewer: What is the purpose of this paragraph? 
It clearly states a size limitation for indexed readers...which are used to overcome the size of big incoming files.
Why do we then mention streaming mode (as opposed to what other mode? Indexed) for even bigger files.
And how is streaming mode limited if it is faster and has no max input size?

Reviewer: Maybe it is my lack of tech knowledge about reading input data, but I am not understanding the differences among....

- "normal" reading of input data
-  indexed readers for large input data (time consuming, space consuming)
-  streaming reading for even larger input data (faster, no memory limit)

The focus should be on defining what indexed readers are, compared to other reading methods (?), and how and why your audience can use them.

> Indexed readers can process input files up to 20GBs. For bigger files use the streaming mode, which is faster. Despite this mode being more limited in functionality, there’s no maximum input size limitation for streaming readers.

Reviewer: 20GBs > 20 GB

> The actual maximum input file size supported is difficult to estimate because it depends on the content of the input.

Reviewer: Does this refer to indexed reader 20-GB size limit?
It's difficult to tell, because it directly follows the comment about streaming not having a size limit....

Reviewer: What do we mean by the content of the input; is this in reference to the following sentence about "different parts" of the input?

> The following list shows the limits on different parts of the input document:

Reviewer: Are these limits that apply to all input? Should it precede the text about using indexed readers, if they are the solution to these limits?
I don't understand the narrative flow of this topic.

* Max nesting depth: `4096`.
* Max individual value size: `4GB`.

> * Tokens: `​​2^32-1` (approximately 4 billion). +

Reviewer: No periods; these are not sentences.

> A token is either a key/value, object-start or array-start, which limits the max document size.

Reviewer: ```suggestion
A token is a key/value pair, an object-start <what? value?>, or an array-start <what?> that limits the max document size.
```

> The amount of the max document size is irrespective of the size of each token in the file. The bigger your average value length, the bigger your file size can be because what is important is the amount of tokens, not their size.

Reviewer: What is the difference between input, document, and input document? Try to standardize the term used to refer to the incoming data.
Ditto maximum input size, maximum input file size, limits on different parts [max] of the input document...and then max nesting, max value size, token limiting max document size.
There seems to be a lot of overlapping similar terminology used in the topic.

Reviewer: ```suggestion
The max document size is based on the number of tokens, not their size. 
```
"The amount of the max document size" > "The max document size"
"The bigger your average value length" > length of what? "The longer (the length)," not "the bigger the length"

> The following examples show input file size samples:

Reviewer: What do these examples have to do with indexed readers?

> In this example, the max file size is approximately 40~50GB:

Reviewer: ```suggestion
In this example, the max file size is between 40 GB and 50 GB:
```

[source,json,linenums]
----
[
  {
    "name": "Mariano",
    "lastName": "Achaval"
  },
  ...
]
----

> The following example is similar to the previous one but minified where the max file size is approximately 30GB:

Reviewer: ```suggestion
In the following example, the max file size is approximately 30 GB:
```

Reviewer: I don't think that "minified" is even a word....

[source,json,linenums]
----
[{"name":"Mariano","lastName":"Achaval"},...]
----

> The following example shows a worst-case scenario where a minified array is number 1s. DataWeave supports only approximately ~8GB of such file:

Reviewer: ```suggestion
The following example shows the worst-case scenario: an array of `1` values. In this case, DataWeave supports approximately 8 GB:
```
